src/main/scala/fi/oph/kouta/repository/toteutusDAO.scala

```diff
@@ -73,17 +71,69 @@ object ToteutusDAO extends ToteutusDAO with ToteutusSQL {
 
   private def insertAsiasanat(toteutus:Toteutus) =
     KeywordDAO.insert(Asiasana, toteutus.metadata.map(_.asiasanat).getOrElse(List()))
+
+  override def getByKoulutusOid(oid: String): List[Toteutus] = {
+    KoutaDatabase.runBlockingTransactionally(
+      for {
+        toteutukset <- selectToteutuksetByKoulutusOid(oid).as[Toteutus]
+        tarjoajat   <- selectToteutustenTarjoajat(toteutukset.map(_.oid.get).toList).as[Tarjoaja]
+      } yield (toteutukset, tarjoajat) ) match {
+        case Left(t) => throw t
+        case Right((toteutukset, tarjoajat)) => {
+          toteutukset.map(t =>
+            t.copy(tarjoajat = tarjoajat.filter(_.oid == t.oid.get).map(_.tarjoajaOid).toList)).toList
+        }
+      }
+  }
+
+  override def listModifiedSince(since:Instant):Seq[String] =
+    KoutaDatabase.runBlocking(selectModifiedSince(since))
 }
 
-sealed trait ToteutusSQL extends ToteutusExtractors with SQLHelpers {
+trait ToteutusModificationSQL extends SQLHelpers {
+  this: ExtractorBase =>
+
+  def selectLastModified(oid:String):DBIO[Option[Instant]] = {
+    sql"""select greatest(
+            max(lower(t.system_time)),
+            max(lower(ta.system_time)),
+            max(upper(th.system_time)),
+            max(upper(tah.system_time)))
+          from toteutukset t
+          left join toteutusten_tarjoajat ta on t.oid = ta.toteutus_oid
+          left join toteutukset_history th on t.oid = th.oid
+          left join toteutusten_tarjoajat_history tah on t.oid = tah.toteutus_oid
+          where t.oid = $oid""".as[Option[Instant]].head
+  }
+
+  def selectModifiedSince(since:Instant): DBIO[Seq[String]] = {
+    sql"""select oid from toteutukset where $since < lower(system_time)
+          union
+          select oid from toteutukset_history where $since <@ system_time
+          union
+          select toteutus_oid from toteutusten_tarjoajat where $since < lower(system_time)
+          union
+          select toteutus_oid from toteutusten_tarjoajat_history where $since <@ system_time""".as[String]
+  }
+
+}
+
+sealed trait ToteutusSQL extends ToteutusExtractors with ToteutusModificationSQL with SQLHelpers {
 
   def selectToteutus(oid:String) =
     sql"""select oid, koulutus_oid, tila, nimi, metadata, muokkaaja, kielivalinta from toteutukset where oid = $oid"""
 
+  def selectToteutuksetByKoulutusOid(oid:String) =
+    sql"""select oid, koulutus_oid, tila, nimi, metadata, muokkaaja, kielivalinta from toteutukset where koulutus_oid = $oid"""
 
   def selectToteutuksenTarjoajat(oid:String) =
     sql"""select toteutus_oid, tarjoaja_oid from toteutusten_tarjoajat where toteutus_oid = $oid"""
 
+  def selectToteutustenTarjoajat(oids:List[String]) = {
+    val toteustusOids = oids.map(s => s"'$s'").mkString(",")
+    sql"""select toteutus_oid, tarjoaja_oid from toteutusten_tarjoajat where toteutus_oid in (#${toteustusOids})"""
```

Comment:
2018-12-03 14:12:25.501 EET [72802] ERROR:  syntax error at or near ")" at character 85
2018-12-03 14:12:25.501 EET [72802] STATEMENT:  select toteutus_oid, tarjoaja_oid from toteutusten_tarjoajat where toteutus_oid in ()

Hajoaa jos toteutusOids tyhjä